File: quota-tray/crates/core/src/aggregate.rs
use crate::types::ProviderSnapshot;

/// `shared_mascot_fill = max(headline_percent)` among healthy providers.
/// If the healthy set is empty, reuse `last_successful` (muted in UI) or `None`.
/// Never invents a live `0.0`.
pub fn aggregate_mascot_fill(
    providers: &[ProviderSnapshot],
    last_successful: Option<f64>,
) -> Option<f64> {
    let mut max_used: Option<f64> = None;
    for snap in providers {
        if !snap.is_healthy() {
            continue;
        }
        let Some(p) = snap.headline_percent else {
            continue;
        };
        max_used = Some(match max_used {
            Some(m) => m.max(p),
            None => p,
        });
    }
    max_used.or(last_successful)
}

File: quota-tray/ui/src/extension/aggregate.ts
import type { ProviderSnapshot } from "../lib/types";

export function isHealthy(snap: ProviderSnapshot): boolean {
  return !snap.stale && snap.error == null && snap.headline_percent != null;
}

/** max(headline_percent) among healthy providers; else lastSuccessful; never a fake 0. */
export function aggregateMascotFill(
  providers: ProviderSnapshot[],
  lastSuccessful: number | null
): number | null {
  let maxUsed: number | null = null;
  for (const snap of providers) {
    if (!isHealthy(snap)) {
      continue;
    }
    if (snap.headline_percent == null) {
      continue;
    }
    maxUsed =
      maxUsed == null
        ? snap.headline_percent
        : Math.max(maxUsed, snap.headline_percent);
  }
  return maxUsed ?? lastSuccessful;
}
